sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp

```diff
@@ -2455,8 +2507,30 @@ ESIMD_INLINE SYCL_ESIMD_FUNCTION __ESIMD_NS::simd<T, N> lsc_load_2d(
   constexpr uint8_t sfid = 0xF;
   constexpr uint8_t numSrc0 = 0x1;
   constexpr uint8_t numDst = (N * sizeof(T)) / 64;
-  return raw_send(oldDst, payload.get_raw_data(), exDesc, desc, execSize, sfid,
-                  numSrc0, numDst);
+  __ESIMD_NS::simd<T, ActualN> Raw =
+      raw_send(oldDst, payload.get_raw_data(), exDesc, desc, execSize, sfid,
+               numSrc0, numDst);
+
+  if constexpr (ActualN == N) {
+    return Raw;
+  } else {
+    // HW restrictions force data which is read to contain padding filled with
+    // garbage for 2d lsc loads. This code eliminates such padding.
+
+    __ESIMD_NS::simd<T, DstElements> Dst;
+
+    for (auto i = 0; i < NBlocks; i++) {
```

Comment:
This code is not super obvious. If you have info about the details how it works, it would be nice to add  a more detailed comment (may be with scheme where garbage-padding is added), while you also remember how it works )

Reply:
It can be a separate PR, if you like.